FEATURE: Deprecate `WorkspaceFinder` and introduce new API

mhsdesign opened the pull request:
introduces

- `ContentRepository::findWorkspaces()`
- `ContentRepository::findWorkspaceByName()`

as well as `find` and `filter` methods on the Workspaces collection to replace all functionality of the `WorkspaceFinder`

This is done as preparation for https://github.com/neos/neos-development-collection/pull/5096 which will remove the Workspace projection and thus the finder.

**Upgrade instructions**

<!-- 
    Add upgrade instructions for breaking changes. 
    Explain who is affected, what has to be adjusted  
-->

**Review instructions**

<!-- 
    If your change is not explained fully by the first section you can 
    add more details here to help the reviewers understand the change.
    We have to understand what you did, why you did it and how we can 
    verify it works correctly and does no harm.
-->

**Checklist**

- [ ] Code follows the PSR-2 coding style
- [ ] Tests have been created, run and adjusted as needed
- [ ] The PR is created against the [lowest maintained branch](https://www.neos.io/features/release-roadmap.html)
- [ ] Reviewer - PR Title is brief but complete and starts with `FEATURE|TASK|BUGFIX`
- [ ] Reviewer - The first section explains the change briefly for change-logs
- [ ] Reviewer - Breaking Changes are marked with `!!!` and have upgrade-instructions

mhsdesign committed:
TASK: Deprecate `WorkspaceFinder` and introduce forward-compatible API

Based on https://github.com/neos/neos-development-collection/pull/5146
Required to ease migration to https://github.com/neos/neos-development-collection/pull/5096

mhsdesign commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -247,6 +244,19 @@ public function getContentGraph(WorkspaceName $workspaceName): ContentGraphInter
+        return $this->getWorkspaceFinder()->findOneByName($workspaceName);
+    }
+
+    public function getWorkspaces(): Workspaces
```
todo discuss `find` vs `get` here. The philosophy to use get was we dont find - eg filter - for anything.
But when we want to use `find` to signal the the implementation is probably slow (at least uncached) then we should go with that.

bwaidelich replied:
I would suggest not to use `find` = `slow` and `get` = `fast` as a _general_ concept, because it tries to change the meaning of those words.
In this case we don't "find" any workspaces, we just get them all – at least that was my original reasoning :)

However we decide, some doc comment would be nice
```suggestion
    /**
     * Returns all workspaces of this content repository. To limit the set, {@see Workspaces::find()} and {@see Workspaces::filter()} can be used
     */
    public function getWorkspaces(): Workspaces
```

nezaniel replied:
iirc our convention is
* get for what is already there (like in-memory computations, property access)
* find for everything that has to be fetched from somewhere

Or in short: yes, we decided to go for get* for fast and find* for slow stuff and I'd rather stick to that.

mhsdesign replied:
Im torn in between. `getContentGraph` for example also does one SQL query to get the currentContentStreamId, though its "get" there ... but im fine with either way.

mhsdesign replied:
I couldnt find discuss post to our find* vs get* philosophy but i found a few old artefacts regarding why `Node::getParent` was renamed to `Node::findParent` in the old Neos world:

> [TODO: rename to findParent() because it is a DB operation](https://github.com/skurfuerst/contentrepository-development-collection/blob/b5e3d80782ff43e686a9001cc7a8f1e82a0dcf32/Neos.EventSourcedContentRepository/Classes/Domain/Projection/Content/TraversableNode.php#L65-L66)

and here https://github.com/neos/neos-development-collection/pull/2202

> adjust TraversableNodeInterface to have proper method namings. Everything which queries something should be called find*; as opposed to get*.

or here https://github.com/neos/neos-development-collection/pull/2202#discussion_r233797157

> `$node->findNodePath()` sounds weird to me. Why not just `getNodePath()`?

> Because we say: "all operations which are slow should be named find*", vs. "all operations which operate on already-loaded data should be named get*". And fetching a node path is a potentially costly operation.

And in `findParent` we also definitely dont filter anything still it continues to be named `Subgraph::findParentNode`.

Though there is one exception `Subgraph::findNodePath` was eventually renamed to `Subgraph::retrieveNodePath` with https://github.com/neos/neos-development-collection/pull/4090
So it seems `retrieveParent` or `retrieveWorkspaces` would also be possibly valid when starting out fresh.

But for now for consistency i guess its a good idea to stick the 6year old philosophy and name it `findWorkspaces` instead (especially we could consider adding a filter later:)).

bwaidelich replied:
Just to avoid confusion: My point was that this kind of "magic meaning" makes sense for a single API, Content(Sub)Graph in this case. I would suggest not to adapt in in _general_ if it leads to weird method names.
But I'm fine with either solution in this case!!

kitsunet replied:
I really find the distinction between `find*` and `get*` hard to grasp, and keep track of. Internals might and can change such that the meaning of find/get is no longer correct, and we cannot change the public API according to our internals. I think this warrants discussion, as I struggled with this naming before as well...

bwaidelich commented on Neos.ContentRepository.Core/Classes/ContentRepository.php:
```diff
@@ -247,6 +244,19 @@ public function getContentGraph(WorkspaceName $workspaceName): ContentGraphInter
         return $this->projectionState(ContentGraphFinder::class)->getByWorkspaceName($workspaceName);
     }
 
+    public function findWorkspaceByName(WorkspaceName $workspaceName): ?Workspace
```
Some doc comment would be nice, e.g:
```suggestion
    /**
     * Returns the workspace with the given name, or NULL if it does not exist in this content repository
     */
    public function findWorkspaceByName(WorkspaceName $workspaceName): ?Workspace
```

bwaidelich reviewed (approved):
Thanks! +1 by reading

dlubitz commented:
Looks good to me, but tests are failing.

mhsdesign committed:
TASK: Update `Workspaces` utilities according to #5096

https://github.com/neos/neos-development-collection/pull/5096/commits/db7861f683f7e308f8515d4be955ccc57dc04fd1

mhsdesign committed:
TASK: Adjust to `find*` vs `get*` convention

> "all operations which are slow should be named find*", vs. "all operations which operate on already-loaded data should be named get*".

mhsdesign commented:
Pushed the comment suggestions on 9.0
